### ISSUE TITLE
Custom properties are missing in the editor 

### ISSUE THREAD

**Issue by benoitchantre** [wprig/wprig#361]
## Issue Overview

Trying to update #332 with the latest changes on `v2.0`, I have noticed that custom properties are not available in the editor even if we import them in `editor-styles.css`. The styles declared in `editor-styles.css` are processed by the block editor and used as inline. We loose our custom properties during that process.

## Possible Solution

Add fallback for custom properties.

**Comment by benoitchantre** [comment]
@ataylorme Were the custom properties fallback removed intentionnaly?
[To support IE11, we need fallbacks](https://caniuse.com/#feat=css-variables) according to caniuse.com

**Comment by ataylorme** [comment]
The variables are imported into the global CSS. I didn't think of the case where editor styles are used stand-alone

> I have noticed that custom properties are not available in the editor even if we import them in editor-styles.css

Are you saying that `@import "../custom-properties.css";` at the top of the `editor-styles.css` files does not work?

If so, that is likely due to [`postcss-import` skipping duplicate imports](https://github.com/postcss/postcss-import#skipduplicates).

cc @mor10

**Comment by benoitchantre** [comment]
> Are you saying that `@import "../custom-properties.css";` at the top of the `editor-styles.css` files does not work?

Custom properties can be found in `editor-styles.min.css`, but this file is then processed by the block editor and not loaded. That process renames some selectors and inlines the styles. This is when we loose our custom properties values.

I think the best solution would be to rename `:root` to `body` in `editor-styles-min.css` (I have tested that manually, it works). An alternative would be to add fallbacks for custom properties.

**Comment by samikeijonen** [comment]
@benoitchantre Did you notice that Core had prefixed `:root` with `.editor-styles-wrapper :root`?

**Comment by mor10** [comment]
This is a bug in the core editor and is going to become an issue down the road. Might be a good cause for a ticket to core or in the Gutenberg repo.

**Comment by mor10** [comment]
@benoitchantre I just tested, and if I add the following to `editor-styles.css`, custom properties are added to the top of the minified editor stylesheet and things work as expected:

```
@import "../custom-media.css";
@import "../custom-properties.css";
```

**Comment by benoitchantre** [comment]
It seems that I haven't been clear enough that this issue has to be seen in the context of #332. Correct me if I'm wrong.

@mor10 I have updated #332 in order to facilitate tests. Can you check again?

> @benoitchantre Did you notice that Core had prefixed `:root` with `.editor-styles-wrapper :root`?

No. How do you see that @samikeijonen? I'm using the inspector and can't find `.editor-styles-wrapper :root`. All styles from the editor stylesheet are inlined.

**Comment by benoitchantre** [comment]
I have now found the styles `.editor-styles-wrapper :root` as mentioned by @samikeijonen.

```
.editor-styles-wrapper :root {
--global-font-color: #333;
--global-font-family: "Crimson Text",serif;
--global-font-size: 20;
--global-font-line-height: 1.4;
--highlight-font-family: "Roboto Condensed",sans-serif;
--content-width: 45rem;
--dropdown-symbol-width: 0.7em;
--color-theme-primary: #e36d60;
--color-theme-secondary: #41848f;
--color-theme-red: #c0392b;
--color-theme-green: #27ae60;
--color-theme-blue: #2980b9;
--color-theme-yellow: #f1c40f;
--color-theme-black: #1c2833;
--color-theme-grey: #95a5a6;
--color-theme-white: #ecf0f1;
--color-custom-daylight: #97c0b7;
--color-custom-sun: #eee9d1;
--color-link: #0073aa;
--color-link-visited: #333;
--color-link-active: #00a0d2;
--font-size-small: calc(16 / var(--global-font-size) * 1rem);
--font-size-regular: calc(var(--global-font-size) / 16 * 1rem);
--font-size-large: calc(36 / var(--global-font-size) * 1rem);
--font-size-larger: calc(48 / var(--global-font-size) * 1rem);
}
```

This is indeed a bug in Core.

**Comment by benoitchantre** [comment]
Related issue in Gutenberg : https://github.com/WordPress/gutenberg/issues/11955

**Comment by ataylorme** [comment]
> until this bug is solved in Core, we can fix that with a string replacement from `:root` to `body` in `editor-styles-min.css`. This can be done by the build tasks.

That is not something trivial to do. 

Will it affect the other styles if we switch from `:root` to `body` in the `custom-properties.css` source file instead?

Are there downsides in switching from `:root` to `body`? I'm not familiar enough with CSS variables to know if it is an okay thing to do or not.

**Comment by benoitchantre** [comment]
I did a test: it works, both on the front-end and in the editor.

Custom properties use cascade. `:root` is similar to `html` with an higher specificy because it's a pseudo element.

Using `body`, we have the following warning during the build:
```
Unexpected duplicate selector "body", first used at line 51   no-duplicate-selectors
```

**Comment by benoitchantre** [comment]
Having fallbacks for custom properties would solve this issue without changing `:root` to `body`.

**Comment by ataylorme** [comment]
Is having fallbacks different than [setting `preserve` to `false`](https://github.com/postcss/postcss-custom-properties#preserve)?

I think a fallback would print both the `var` and the computed values where `preserve` being `false` would just compute the value and remove `var` statements from the minified CSS.

As I said, I am not much of a front-end expert but happy to adjust the gulp settings/change the build process if there is a consensus on a better approach.

**Comment by benoitchantre** [comment]
> Is having fallbacks different than [setting `preserve` to `false`](https://github.com/postcss/postcss-custom-properties#preserve)?
> 
> I think a fallback would print both the `var` and the computed values where `preserve` being `false` would just compute the value and remove `var` statements from the minified CSS.

Correct.

I have tried to set `preserve` to `false` in the styles task, but haven't notice a replacement of `var` by the computed value. Can you look at that?

**Comment by benoitchantre** [comment]
> I have tried to set preserve to false in the styles task, but haven't notice a replacement of var by the computed value. Can you look at that?

@ataylorme I have found why setting preserve to `false` doesn't work and can provide a fix. Should I create a separate issue and pull request or can I fix that in #332?

**Comment by ataylorme** [comment]
Please make a new issue and PR to keep things clean. I know Morten has some opinions on why preserve should be true so we can discuss there.

Once it is merged we can update this PR with the fix.

**Comment by mor10** [comment]
See https://github.com/WordPress/gutenberg/issues/14297 and https://github.com/WordPress/gutenberg/issues/11955 for further detail on why this happens.

**Comment by mor10** [comment]
I also made a Trac ticket for this as it's already live in core: https://core.trac.wordpress.org/ticket/46435#ticket

**Comment by ataylorme** [comment]
Moving conversation from #332 here.

> Would it be complicated to have in the build task a separated setting (or hard coded value) for the editor stylesheet, with a default value for `preserve` set to `false`?

Yes! See #369.

> There is a plugin to swap `:root` for `html`. `jonathantneal/postcss-unroot`

Thank you @jonathantneal! I used that in #370.

We now have 3 pull request for options to address this issue

* #368 - Automatically transform editor CSS with custom media and custom properties inline during the build process.
* #369 - Dynamically set postcss `preserve` to `false` for editor CSS.
* #370 - New editor styles task that uses `postcss-unroot`.

Personally, I prefer #369. Yes, we are tapping into the CSS files in the gulp stream but I see this as a temporary solution. #370, creating a whole new task, changes a lot more. If we needed to do something different with editor styles long-term I would be okay with that but since we will revert this workaround when WordPress core is fixed, I think a bit of a hard-coded solution short-term is okay.

**Comment by benoitchantre** [comment]
Wow. Thank you for all your work @ataylorme.

I personally prefer #369 for the same reasons.

**Comment by benoitchantre** [comment]
Custom properties under `:root` will work with WordPress 5.2.
https://github.com/WordPress/gutenberg/issues/14297#issuecomment-470511862

**Comment by ataylorme** [comment]
@mor10 are you okay with #369, dynamically set postcss `preserve` to `false` for editor CSS, as a solution?

It can be reverted once WordPress `5.2` is released, which is currently planned for April 30th.
